src/psyclone/tests/psyir/backend/fortran_test.py

```diff
@@ -0,0 +1,589 @@
+    result = fvisitor.get_declaration(symbol)
+    assert result == "integer(i_def), dimension(2,:,2) :: dummy2\n"
+
+    # Constant
+    symbol = Symbol("dummy3", "integer", constant_value=10)
+    result = fvisitor.get_declaration(symbol)
+    assert result == "integer(i_def), parameter :: dummy3 = 10\n"
+
+
+def create_schedule(code):
+    '''Utility function that returns a PSyIR tree from Fortran
+    code using fparser2 and Fparser2ASTProcessor.
+
+    :param str code: Fortran code.
+
+    :returns: PSyIR tree representing the Fortran code.
+    :rtype: Subclass of :py:class:`psyclone.psyGen.Node`
+
+    '''
+    reader = FortranStringReader(code)
+    f2003_parser = ParserFactory().create(std="f2003")
+    parse_tree = f2003_parser(reader)
+
+    # Generate PSyIR schedule from fparser2 parse tree
+    processor = Fparser2ASTProcessor()
+    schedule = processor.generate_schedule("tmp", parse_tree)
+
+    return schedule
+
+
+def test_FortranWriter_exception():
+    '''Check the FortranWriter class instance raises an exception if an
+    unsupported PSyIR node is found.
+
+    '''
+    # Generate fparser2 parse tree from Fortran code.
+    code = (
+        "module test\n"
+        "contains\n"
+        "subroutine tmp()\n"
+        "  integer :: a,b,c\n"
+        "  a = b/c\n"
+        "end subroutine tmp\n"
+        "end module test")
+    schedule = create_schedule(code)
+
+    # modify the reference to b to be something unsupported
+    class Unsupported(Node):
+        '''A PSyIR node that will not be supported by the Fortran visitor.'''
+
+    unsupported = Unsupported()
+    assignment = schedule.children[0]
+    binary_operation = assignment.children[1]
+    assignment.children[1] = unsupported
+    unsupported.children = binary_operation.children
+
+    # Generate Fortran from the PSyIR schedule
+    fvisitor = FortranWriter()
+    with pytest.raises(VisitorError) as excinfo:
+        result = fvisitor(schedule)
+    assert "Unsupported node 'Unsupported' found" in str(excinfo)
+
+
+def test_FortranWriter_nemokern():
+    '''Check the FortranWriter class nemokern method prints the
+    class information and calls any children. This method is used to
+    output nothing for a NemoKern object and simply call its children
+    as NemoKern is a collection of PSyIR nodes so needs no
+    output itself.
+
+    '''
+    # Generate fparser2 parse tree from Fortran code.
+    code = (
+        "module test\n"
+        "contains\n"
+        "subroutine tmp()\n"
+        "  integer :: a,b,c\n"
+        "  a = b/c\n"
+        "end subroutine tmp\n"
+        "end module test")
+    schedule = create_schedule(code)
+
+    # add a NemoKern object to the tree
+    from psyclone.nemo import NemoKern
+    nemo_kern = NemoKern(schedule.children, None, parent=schedule)
+    schedule.children = [nemo_kern]
+
+    # Generate Fortran from the PSyIR schedule
+    fvisitor = FortranWriter()
+    result = fvisitor.visit(schedule)
```

Comment:
There are some tests with the `visit()` method and others with the function call, I recommend just using the function call.

Reply:
Changed.